### PR TITLE
Add the support for on-demand mode

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Context
As per https://github.com/nvaccess/nvda/pull/15804, a new on-demand speech mode has been implemented in NVDA.
This speech mode is like speech mode off, except for commands that perform no other action than reporting a piece of information.

### Description of development approach
All the scripts that only report information are included in the on-demand speech mode. These are all the scripts of this add-on, except the 3 ones which perform an action:
- swap language
- copy last translation
- open translation settings dialog

### Description of development approach
This add-on is a good example for such implementation since various issues have been solved to include the on-demand mode for its scripts.

#### Standard approach:
The `speakOnDemand` parameter has been passed to the target scripts so that these scripts still speak in on-demand mode. This is the normal procedure.

#### Modify the `finally_` function
The `finally_` function used for layered gesture has been modified. `finally_(func, final)` returns a function that calls `final` after `func`, even if it fails. However, this new function was wrapped with the properties of `final`. This has no sense and when `func` is a script, it prevented this new function to have the same properties.
Now, I have wrapped this new function with the properties of `func` so that the `speakOnDemand` parameter is honoured even for layered scripts.

Cc @tspivey and @josephsl for double check / validation, since you are the ones who have worked on the `finally_` function.

#### Support delayed functions
The on-demand parameter only impacts speech performed in the flagged script. However some script schedule `speak` function calls after the script is finished, e.g. because they launch a thread and want the result to be spoken once the thread has finished its work.

Thus I have created a specific `funcExecutor` module with a `executeWithSpeakOnDemand` function. This function allow to call a function forcing the "on-demand" mode for its execution, even if this function executes out of any script.

### Testing
Tested with NVDA 2019.3.1, NVDA 2023.3 and last beta pre-2024.1 (commit b93e08f308d149f78b1041416240397d522816c0).

### CODE REVIEW

**Comment by josephsl** [comment]
Hi,

Does this work in NVDA 2023.3 i.e. care is taken to define speak on demand argument if NVDA 2024.1 or later is in use?

Thanks.

**Comment by CyrilleB79** [comment]
> Does this work in NVDA 2023.3 i.e. care is taken to define speak on demand argument if NVDA 2024.1 or later is in use?

As written in the initial description of this PR, the add-on has been tested on NVDA 2019.3.1, 2023.3 and 2024.1.
This add-on still uses the old script definition with attributes (no `@script `decorator). So there is no issue with unsupported parameter for the `@script` decorator.

@beqabeqa473, now that this add-on is only compatible with NVDA 2019.3 onward, I may convert to `@cript` syntax while at it. Just let me know.

In this case, I can use a code like the following:
```
try:
    # NVDA >= 2024.1
    speech.speech.SpeechMode.onDemand
    speakOnDemand = {'speakOnDemand': True}
except AttributeError:
    # NVDA <= 2023.3
    speakOnDemand = {}

@script(
    description="foo",
    gesture='kb:NVDA+F6',
    **speakOnDemand,
)
def script_foo(self, g):
    ui.message('Hello world!')
```

Note for @josephsl: I have tagged you especially for your review on the modifications in the `finally_` function. Of course comments on other parts of the code are welcome too.

**Comment by josephsl** [comment]
Hi, actual credit should go to Tyler. Thanks.

**Comment by beqabeqa473** [comment]
hi @CyrilleB79.

Thank you for your work.
Yes, ket's rewrite using script decorator.

**Comment by CyrilleB79** [comment]
Thanks, I'll convert to script decorator.

I realize that you have merged #30 into stable, not into master. I do not know if it's because I have targeted stable branch myself or if it's because you have changed the target branch before merging.

The current PR, instead, is against master. Should I re-target against stabl as #30, or should I keep master branch as target?

**Comment by beqabeqa473** [comment]
Seems you targetted it to stable.

On 12/17/23, Cyrille Bougot ***@***.***> wrote:
> Thanks, I'll convert to script decorator.
>
> I realize that you have merged #30 into stable, not into master. I do not
> know if it's because I have targeted stable branch myself or if it's because
> you have changed the target branch before merging.
>
> The current PR, instead, is against master. Should I re-target against stabl
> as #30, or should I keep master branch as target?
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/beqabeqa473/instantTranslate/pull/31#issuecomment-1859146358
> You are receiving this because you were mentioned.
>
> Message ID: ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by CyrilleB79** [comment]
OK, I have pushed the conversion to script decorator.

I have also realized that I had forgotten to push some changes in 
`funcExecutor.py` (hence Joseph's question on the versions on which it should work or not). It is now fixed and it should work with any version of NVDA.

@beqabeqa473, let me know if you plan to merge stable to master and/or if I have to rebase this PR to target stable.

Note that the default branch for your repository is stable, not master as it is usually. That's probably why I have targeted this branch for PR #30, without paying attention to it.

**Comment by beqabeqa473** [comment]
Don't worry. I will synchronize it. I have also some undergoing work
for this addon, so will wait to 2024.1 before release.

On 12/17/23, Cyrille Bougot ***@***.***> wrote:
> OK, I have pushed the conversion to script decorator.
>
> I have also realized that I had forgotten to push some changes in
> `funcExecutor.py` (hence Joseph's question on the versions on which it
> should work or not). It is now fixed and it should work with any version of
> NVDA.
>
> @beqabeqa473, let me know if you plan to merge stable to master and/or if I
> have to rebase this PR to target stable.
>
> Note that the default branch for your repository is stable, not master as it
> is usually. That's probably why I have targeted this branch for PR #30,
> without paying attention to it.
>
>
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/beqabeqa473/instantTranslate/pull/31#issuecomment-1859196060
> You are receiving this because you were mentioned.
>
> Message ID: ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by CyrilleB79** [comment]
@beqabeqa473 since you have not yet merged this PR, I plan to rewrite totally `executeWithSpeakOnDemand` function, according to a discussion with @paulber19 and @abdel792. The new function should be much simpler, based on a very interesting code suggestion from @paulber19.

In the case of the current version of this add-on, it should make no difference in the result. However, the new pattern should prevent potential issues in the future in case you plan to use it with scripts using double or triple press.

Until the new version of this function is ready, I convert this PR to draft.

**Comment by beqabeqa473** [comment]
Ok. I will merge this when NVDA 2024.1 will be released.

On 1/7/24, Cyrille Bougot ***@***.***> wrote:
> @beqabeqa473 since you have not yet merged this PR, I plan to rewrite
> totally `executeWithSpeakOnDemand` function, according to a discussion with
> @paulber19 and @abdel792. The new function should be much simpler, based on
> a very interesting code suggestion from @paulber19.
>
> In the case of the current version of this add-on, it should make no
> difference in the result. However, the new pattern should prevent potential
> issues in the future in case you plan to use it with scripts using double or
> triple press.
>
> Until the new version of this function is ready, I convert this PR to
> draft.
>
>
>
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/beqabeqa473/instantTranslate/pull/31#issuecomment-1879872540
> You are receiving this because you were mentioned.
>
> Message ID: ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by CyrilleB79** [comment]
The PR is now ready again and tested with NVDA 2024.1beta3 as well as with NVDA pre-2024.1 (actually 2023.2).

I have gathered all the useful functions related to on-demand speech mode support in the same module that can be used by other add-ons as well. The new module does not create a fake script anymore since it was causing issue with the `scriptHandler.getLastScriptRepeatCount` function to detect multi-presses. It just temporarily passes the speech mode to Talk during the execution of a function that need to speak in the on-demand mode. To avoid potential mismatch in the speech mode state, I have restricted the execution of `executeWithSpeakOnDemand` to the main thread.

File: addon/globalPlugins/instantTranslate/speechOnDemand.py
Comment (lukaszgo1) [inline addon/globalPlugins/instantTranslate/speechOnDemand.py]: `isMainThread` was introduced around 2023, so it won't work for older versions of NVDA. `core.mainThreadId` should be used here.

**Review by lukaszgo1 — CHANGES_REQUESTED** [review]
Thanks for writing this reference implementation - this will certainly be useful for many add-on authors. My only comment is a small nit  regarding backward compatibility.

File: addon/globalPlugins/instantTranslate/speechOnDemand.py
Comment (CyrilleB79) [reply addon/globalPlugins/instantTranslate/speechOnDemand.py]: Many thanks @lukaszgo1 for your review!

After doing changes, I had not tested again against various versions.
It is now tested with NVDA 2019.3.1 and NVDA 2024.1beta3.

**Comment by CyrilleB79** [comment]
@beqabeqa473 thanks for having just merged my two last PRs.

For future contributions, could you clarify if you expect to have PRs done against master, what is usual with NVDA add-ons, or against stable branch?

If you expect against master, could you configure the default branch of this repository as such in its settings? Just ask for help if needed.

**Pull request merged**
